### DISCUSSION TITLE
[Bug]: Undefined Symbol for architecture arm 64

### DISCUSSION THREAD

**Discussion by MikePendo** [theos/theos#677] [Questions]
### What are the steps to reproduce this issue?

My goal is  embed my custom framework with tweak to the deb file and sidelod it to  the ipa later on.
MakeFile:
```
TARGET := iphone:clang:latest:14.4
ARCHS = arm64 arm64e
GO_EASY_ON_ME = 1
FINALPACKAGE = 1
DEBUG = 0
include $(THEOS)/makefiles/common.mk
TWEAK_NAME = SiteReconTweak
SiteReconTweak_FILES = Tweak.x
SiteReconTweak_CFLAGS =  -fobjc-arc -I./MyFramework.framework/Headers
SiteReconTweak_LDFLAGS = -L./MyFramework.framework 
include $(THEOS_MAKE_PATH)/tweak.mk
```
Tweak.x
```
#import "MyFrameworkManager.h"
#import <UIKit/UIKit.h>

%hook UIWindow
- (void)motionEnded:(UIEventSubtype)motion withEvent:(UIEvent *)event {
  %orig;
  [[MyFrameworkManager sharedManager] setup:@""];
}
```

The whole project:
<img width="705" alt="Screen Shot 2022-07-24 at 10 46 34" src="https://user-images.githubusercontent.com/56674958/180637527-e41e86df-e791-47dd-9a4d-1fd5d8546764.png">
the framework supports the following archs:
```
MyFramework (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
MyFramework (for architecture i386):	Mach-O dynamically linked shared library i386
MyFramework (for architecture armv7):	Mach-O dynamically linked shared library arm_v7
MyFramework (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
```
I have found the following issues which are similar to mine:
[How to include 3rd Party Framework?](https://github.com/theos/theos/issues/124)
[My tweak stops working after adding a custom dylib file to it](https://github.com/theos/theos/issues/256)

current error:
```
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_MyFrameworkManager", referenced from:
      objc-class-ref in Tweak.x.fafea5e8.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I can import the `.h` but that fail to link any way. 

I am aware that embed the framework in such way is not a best practice BUT it still should be possible.
I am also not 100% if my flags in the makefile are correct and where can I find a proper docs about makefile flags.
Can somebody assist me in that matter? or maybe the slicing is not well supported?
Thanks in advance 

### What happens?

make package 
```
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_MyFrameworkManager", referenced from:
      objc-class-ref in Tweak.x.fafea5e8.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### What were you expecting to happen?

successfully embeds my framework in .deb

### Paste any relevant logs, error output, etc.

```shell
make package
==> Notice: Build may be slow as Theos isn’t using all available CPU cores on this computer. 
Consider upgrading GNU Make: https://theos.dev/docs/parallel-building
> Making all for tweak SiteReconTweak…
==> Preprocessing Tweak.x…
==> Compiling Tweak.x (arm64)…
==> Linking tweak SiteReconTweak (arm64)…
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_MyFrameworkManager", referenced from:
      objc-class-ref in Tweak.x.fafea5e8.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [theos/siterecontweak/.theos/obj/arm64/SiteReconTweak.dylib] Error 1
make[2]: *** [theos/siterecontweak/.theos/obj/arm64/SiteReconTweak.dylib] Error 2
make[1]: *** [internal-library-all_] Error 2
make: *** [SiteReconTweak.all.tweak.variables] Error 2
```

### What OS are you seeing the problem on?

macOS

### What OS version does it have installed?

Monterey

### What toolchain and version are you using?

Xcode 13.4

### Which SDK version are you using?

iOS 15

### What OS is your client device running?

iOS

### What OS version does it have installed?

io6

### Which device model is it?

iPhone 13 mini

### Which jailbreak is it using (if relevant)?

no

**Comment by L1ghtmann** [comment]
So a few things:

1) -L specifies an additional *library* search path, you want -F to specify an additional *framework* search path;

2) -F expects a directory with .frameworks, not a path to a specific .framework, so you'll need to place the .framework in a directory (e.g., ./Frameworks/). That would change the additional include path to ```-I./Frameworks/MyFramework.framework/Headers/``` and the additional framework path to ```-F./Frameworks/```;

3) To fix the undefined symbols issue, you'll want to tell the linker to link the framework. To do that, you'll want to add ```SiteReconTweak_EXTRA_FRAMEWORKS = MyFramework``` to your Makefile as well (this just adds ```-framework MyFramework``` to the linker flags for you)


>  I am also not 100% if my flags in the makefile are correct and where can I find a proper docs about makefile flags.

Clang's [documentation](https://clang.llvm.org/docs/CommandGuide/clang.html) and ld's [documentation](https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_node/ld_3.html) are both super helpful when you want to know which flag(s) to use.
